File: src/main/java/com/regionrental/managers/SignManager.java
package com.regionrental.managers;

import com.regionrental.RegionRental;
import com.regionrental.util.WorldRegionParser;
import org.bukkit.ChatColor;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.Sign;
import org.bukkit.block.sign.Side;
import org.bukkit.block.data.BlockData;
import org.bukkit.block.data.type.WallSign;

import java.util.HashSet;
import java.util.Map;
import java.util.Set;

public class SignManager {

    private final RegionRental plugin;
    private final Set<String> dirtySigns;  // Track signs that need updates

    public SignManager(RegionRental plugin) {
        this.plugin = plugin;
        this.dirtySigns = new HashSet<>();
    }
    
    public void loadAllSigns() {
        // Migrate existing signs to include support block data
        migrateSupportBlocks();

        // Mark all signs as dirty to update them on startup
        markAllSignsDirty();

        // Update all dirty signs
        updateAllSigns();
    }

    /**
     * Marks all signs as dirty for initial update or full refresh.
     */
    private void markAllSignsDirty() {
        Map<String, Location> signs = plugin.getSignsConfig().getAllSigns();
        dirtySigns.addAll(signs.keySet());
    }

    /**
     * Migrates existing signs to include support block data
     * This is for backward compatibility with signs created before this feature
     */
    private void migrateSupportBlocks() {
        Map<String, Location> signs = plugin.getSignsConfig().getAllSigns();
        int migratedCount = 0;

        for (String compositeKey : signs.keySet()) {
            // Parse composite key "world:region"
            String worldName = WorldRegionParser.extractWorldName(compositeKey);
            String regionName = WorldRegionParser.extractRegionName(compositeKey);

            if (worldName == null) {
                plugin.getLogger().warning("Invalid composite key format: " + compositeKey);
                continue;
            }

            World world = plugin.getServer().getWorld(worldName);
            if (world == null) {
                plugin.getLogger().warning("World not found for sign: " + worldName);
                continue;
            }

            // Skip if support block already exists
            if (plugin.getSignsConfig().hasSupportBlock(regionName, world)) {
                continue;
            }

            Location signLoc = signs.get(compositeKey);
            if (signLoc == null) {
                continue;
            }

            Block signBlock = signLoc.getBlock();
            if (!(signBlock.getState() instanceof Sign)) {
                plugin.getLogger().warning("Sign for region " + compositeKey + " no longer exists at location");
                continue;
            }

            // Detect support block
            Block supportBlock = getSupportBlock(signBlock);
            if (supportBlock != null) {
                // Store the support block information
                Location supportLoc = supportBlock.getLocation();
                String blockType = supportBlock.getType().name();
                String blockData = supportBlock.getBlockData().getAsString();

                plugin.getSignsConfig().addSupportBlock(regionName, world, supportLoc, blockType, blockData);
                migratedCount++;

                plugin.getLogger().info("Migrated support block for region " + compositeKey +
                    " (Type: " + blockType + ")");
            } else {
                plugin.getLogger().warning("Could not detect support block for existing sign in region " + compositeKey);
            }
        }

        if (migratedCount > 0) {
            plugin.getLogger().info("Migrated " + migratedCount + " sign(s) with support block protection");
        }
    }
    
    public void createSign(String regionName, Location location) {
        // Store sign location (addSign uses location's world automatically)
        plugin.getSignsConfig().addSign(regionName, location);

        // Detect and store support block
        Block signBlock = location.getBlock();
        Block supportBlock = getSupportBlock(signBlock);

        if (supportBlock != null) {
            // Store the support block information
            Location supportLoc = supportBlock.getLocation();
            String blockType = supportBlock.getType().name();
            String blockData = supportBlock.getBlockData().getAsString();

            plugin.getSignsConfig().addSupportBlock(regionName, location.getWorld(), supportLoc, blockType, blockData);
            plugin.getLogger().info("Stored support block for region " + location.getWorld().getName() + ":" + regionName +
                " (Type: " + blockType + " at " + supportLoc.getBlockX() + "," +
                supportLoc.getBlockY() + "," + supportLoc.getBlockZ() + ")");
        } else {
            plugin.getLogger().warning("Could not detect support block for sign at " + location);
        }

        // Update the sign
        updateSign(regionName, location.getWorld());

        plugin.getLogger().info("Created rental sign for region " + location.getWorld().getName() + ":" + regionName);
    }

    /**
     * Gets the support block for a sign (the block it's attached to or placed on)
     * @param signBlock The sign block
     * @return The support block, or null if not found
     */
    private Block getSupportBlock(Block signBlock) {
        BlockData blockData = signBlock.getBlockData();

        // Check if it's a wall sign
        if (blockData instanceof WallSign) {
            WallSign wallSign = (WallSign) blockData;
            BlockFace facing = wallSign.getFacing();
            // Wall signs are attached to the block opposite to their facing direction
            return signBlock.getRelative(facing.getOppositeFace());
        }
        // Otherwise it's a standing sign - support block is below
        else if (signBlock.getType().name().contains("SIGN")) {
            return signBlock.getRelative(BlockFace.DOWN);
        }

        return null;
    }
    
    public void removeSign(String regionName, World world) {
        plugin.getSignsConfig().removeSign(regionName, world);
        plugin.getLogger().info("Removed rental sign for region " + world.getName() + ":" + regionName);
    }

    /**
     * Completely removes RegionRental setup from a region
     * This includes removing the sign from config and restoring the support block
     * @param regionName The region to remove setup from
     * @param world The world the region is in
     * @return true if sign was removed, false if no sign existed
     */
    public boolean removeRegionSetup(String regionName, World world) {
        Location location = plugin.getSignsConfig().getSignLocation(regionName, world);

        if (location == null) {
            return false; // No sign exists for this region
        }

        // Restore the support block if it exists
        if (plugin.getSignsConfig().hasSupportBlock(regionName, world)) {
            Location supportLoc = plugin.getSignsConfig().getSupportBlockLocation(regionName, world);
            Map<String, String> supportData = plugin.getSignsConfig().getSupportBlockData(regionName, world);

            if (supportLoc != null && supportData != null) {
                try {
                    Block supportBlock = supportLoc.getBlock();
                    Material originalType = Material.valueOf(supportData.get("type"));
                    String originalData = supportData.get("data");

                    // Restore the block type
                    supportBlock.setType(originalType);

                    // Restore the block data if it exists
                    if (originalData != null && !originalData.isEmpty()) {
                        try {
                            BlockData blockData = plugin.getServer().createBlockData(originalData);
                            supportBlock.setBlockData(blockData);
                        } catch (IllegalArgumentException e) {
                            plugin.getLogger().warning("Could not restore block data for support block: " + e.getMessage());
                        }
                    }

                    plugin.getLogger().info("Restored support block for region " + world.getName() + ":" + regionName +
                        " to " + originalType.name());
                } catch (IllegalArgumentException e) {
                    plugin.getLogger().warning("Could not restore support block for region " + world.getName() + ":" + regionName +
                        ": Invalid material type - " + e.getMessage());
                }
            }
        } else {
            // No support block data - manually break the sign
            Block block = location.getBlock();
            if (block.getState() instanceof Sign) {
                block.setType(Material.AIR);
            }
        }

        // Remove from config (this removes both sign and support block data)
        plugin.getSignsConfig().removeSign(regionName, world);
        plugin.getLogger().info("Removed RegionRental setup from region " + world.getName() + ":" + regionName);

        return true;
    }
    
    public void updateSign(String regionName, org.bukkit.World world) {
        Location location = plugin.getSignsConfig().getSignLocation(regionName, world);

        if (location == null) {
            return;
        }

        Block block = location.getBlock();

        // Check if block is a sign
        if (!(block.getState() instanceof Sign)) {
            plugin.getLogger().warning("Sign location for " + world.getName() + ":" + regionName + " is not a sign!");
            return;
        }

        Sign sign = (Sign) block.getState();
        Rental rental = plugin.getRentalManager().getRental(regionName, world);

        if (rental == null) {
            // Region is available
            updateAvailableSign(sign, regionName);
        } else {
            // Region is rented
            updateRentedSign(sign, rental);
        }

        sign.update(true);
    }

    // Backward compatibility - deprecated, use world-aware version
    @Deprecated
    public void updateSign(String regionName) {
        // Note: This is deprecated because it doesn't know which world to check
        // It will attempt to find the sign but may not work correctly with multiple worlds
        plugin.getLogger().warning("Deprecated updateSign(regionName) called for: " + regionName +
            " - Please update to use updateSign(regionName, world)");
    }
    
    private void updateAvailableSign(Sign sign, String regionName) {
        World world = sign.getWorld();
        double price = plugin.getConfigManager().getPriceForRegion(regionName, world);
        int duration = plugin.getConfigManager().getDurationForRegion(regionName, world);
        String formattedPrice = String.format(plugin.getConfigManager().getCurrencyFormat(), price);

        int line = 0;
        for (String format : plugin.getConfigManager().getAvailableSignFormat()) {
            if (line >= 4) break;

            String text = format
                .replace("{region}", regionName)
                .replace("{price}", formattedPrice)
                .replace("{duration}", String.valueOf(duration));

            sign.line(line, net.kyori.adventure.text.Component.text(color(text)));
            line++;
        }
    }
    
    private void updateRentedSign(Sign sign, Rental rental) {
        int line = 0;
        for (String format : plugin.getConfigManager().getRentedSignFormat()) {
            if (line >= 4) break;
            
            String text = format
                .replace("{region}", rental.getRegionName())
                .replace("{owner}", rental.getPlayerName())
                .replace("{expires}", rental.getFormattedEndDate())
                .replace("{days}", String.valueOf(rental.getDaysRemaining()))
                .replace("{hours}", String.valueOf(rental.getHoursRemaining()));
            
            sign.line(line, net.kyori.adventure.text.Component.text(color(text)));
            line++;
        }
    }
    
    /**
     * Marks a sign as dirty (needs update).
     * Uses composite key format "world:region".
     *
     * @param regionName The region name
     * @param world The world
     */
    public void markSignDirty(String regionName, World world) {
        String compositeKey = world.getName() + ":" + regionName;
        dirtySigns.add(compositeKey);
    }

    /**
     * Updates all dirty signs and clears the dirty set.
     * This is much more efficient than updating ALL signs every 30 seconds.
     */
    public void updateAllSigns() {
        if (dirtySigns.isEmpty()) {
            return;  // Nothing to update
        }

        Map<String, Location> allSigns = plugin.getSignsConfig().getAllSigns();

        // Update only dirty signs
        for (String compositeKey : dirtySigns) {
            Location location = allSigns.get(compositeKey);
            if (location == null) {
                continue;  // Sign was removed
            }

            // Parse composite key "world:region"
            String regionName = WorldRegionParser.extractRegionName(compositeKey);
            World world = location.getWorld();

            if (world != null) {
                updateSign(regionName, world);
            }
        }

        // Clear dirty set after updating
        dirtySigns.clear();
    }
    
    public boolean isRentalSign(Location location) {
        return plugin.getSignsConfig().getRegionByLocation(location) != null;
    }
    
    public String getRegionFromSign(Location location) {
        return plugin.getSignsConfig().getRegionByLocation(location);
    }

    /**
     * Gets the region name if the given location is a protected support block
     * @param location The block location to check
     * @return Composite key "world:region" if this is a support block, null otherwise
     */
    public String getSupportBlockRegion(Location location) {
        Map<String, Location> signs = plugin.getSignsConfig().getAllSigns();

        for (String compositeKey : signs.keySet()) {
            // Parse composite key "world:region"
            String worldName = WorldRegionParser.extractWorldName(compositeKey);
            String regionName = WorldRegionParser.extractRegionName(compositeKey);

            if (worldName == null) {
                continue;
            }

            World world = plugin.getServer().getWorld(worldName);
            if (world == null) {
                continue;
            }

            Location supportLoc = plugin.getSignsConfig().getSupportBlockLocation(regionName, world);
            if (supportLoc != null && isSameLocation(location, supportLoc)) {
                return compositeKey; // Return full composite key
            }
        }

        return null;
    }

    /**
     * Checks if two locations represent the same block
     */
    private boolean isSameLocation(Location loc1, Location loc2) {
        return loc1.getWorld().getName().equals(loc2.getWorld().getName()) &&
               loc1.getBlockX() == loc2.getBlockX() &&
               loc1.getBlockY() == loc2.getBlockY() &&
               loc1.getBlockZ() == loc2.getBlockZ();
    }

    private String color(String text) {
        return ChatColor.translateAlternateColorCodes('&', text);
    }
}

File: src/main/kotlin/com/zonerental/managers/SignManager.kt
package com.zonerental.managers

import com.zonerental.ZoneRental
import com.zonerental.extensions.toComponent
import com.zonerental.util.WorldRegionParser
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.World
import org.bukkit.block.Block
import org.bukkit.block.BlockFace
import org.bukkit.block.Sign
import org.bukkit.block.data.type.WallSign

/**
 * Manages rental signs for regions.
 */
class SignManager(private val plugin: ZoneRental) {

    private val dirtySigns = mutableSetOf<String>() // Track signs that need updates

    fun loadAllSigns() {
        // Migrate existing signs to include support block data
        migrateSupportBlocks()

        // Mark all signs as dirty to update them on startup
        markAllSignsDirty()

        // Update all dirty signs
        updateAllSigns()
    }

    /**
     * Marks all signs as dirty for initial update or full refresh.
     */
    private fun markAllSignsDirty() {
        val signs = plugin.signsConfig.getAllSigns()
        dirtySigns.addAll(signs.keys)
    }

    /**
     * Migrates existing signs to include support block data.
     * This is for backward compatibility with signs created before this feature.
     */
    private fun migrateSupportBlocks() {
        val signs = plugin.signsConfig.getAllSigns()
        var migratedCount = 0

        for ((compositeKey, signLoc) in signs) {
            // Parse composite key "world:region"
            val worldName = WorldRegionParser.extractWorldName(compositeKey)
            val regionName = WorldRegionParser.extractRegionName(compositeKey)

            if (worldName == null || regionName == null) {
                plugin.logger.warning("Invalid composite key format: $compositeKey")
                continue
            }

            val world = plugin.server.getWorld(worldName)
            if (world == null) {
                plugin.logger.warning("World not found for sign: $worldName")
                continue
            }

            // Skip if support block already exists
            if (plugin.signsConfig.hasSupportBlock(regionName, world)) {
                continue
            }

            val signBlock = signLoc.getBlock()

            if (signBlock.state !is Sign) {
                plugin.logger.warning("Sign for region $compositeKey no longer exists at location")
                continue
            }

            // Detect support block
            val supportBlock = getSupportBlock(signBlock)
            if (supportBlock != null) {
                // Store the support block information
                val supportLoc = supportBlock.location
                val blockType = supportBlock.type.name
                val blockData = supportBlock.blockData.asString

                plugin.signsConfig.addSupportBlock(regionName, world, supportLoc, blockType, blockData)
                migratedCount++

                plugin.logger.info("Migrated support block for region $compositeKey (Type: $blockType)")
            } else {
                plugin.logger.warning("Could not detect support block for existing sign in region $compositeKey")
            }
        }

        if (migratedCount > 0) {
            plugin.logger.info("Migrated $migratedCount sign(s) with support block protection")
        }
    }

    fun createSign(regionName: String, location: Location) {
        // Store sign location (addSign uses location's world automatically)
        plugin.signsConfig.addSign(regionName, location)

        // Detect and store support block
        val signBlock = location.getBlock()
        val supportBlock = getSupportBlock(signBlock)

        if (supportBlock != null) {
            // Store the support block information
            val supportLoc = supportBlock.location
            val blockType = supportBlock.type.name
            val blockData = supportBlock.blockData.asString

            plugin.signsConfig.addSupportBlock(regionName, location.world, supportLoc, blockType, blockData)
            plugin.logger.info(
                "Stored support block for region ${location.world.name}:$regionName " +
                "(Type: $blockType at ${supportLoc.blockX},${supportLoc.blockY},${supportLoc.blockZ})"
            )
        } else {
            plugin.logger.warning("Could not detect support block for sign at $location")
        }

        // Update the sign
        updateSign(regionName, location.world)

        plugin.logger.info("Created rental sign for region ${location.world.name}:$regionName")
    }

    /**
     * Gets the support block for a sign (the block it's attached to or placed on).
     */
    private fun getSupportBlock(signBlock: Block): Block? {
        val blockData = signBlock.blockData

        // Check if it's a wall sign
        if (blockData is WallSign) {
            val facing = blockData.facing
            // Wall signs are attached to the block opposite to their facing direction
            return signBlock.getRelative(facing.oppositeFace)
        }

        // Otherwise it's a standing sign - support block is below
        if (signBlock.type.name.contains("SIGN")) {
            return signBlock.getRelative(BlockFace.DOWN)
        }

        return null
    }

    fun removeSign(regionName: String, world: World) {
        plugin.signsConfig.removeSign(regionName, world)
        plugin.logger.info("Removed rental sign for region ${world.name}:$regionName")
    }

    /**
     * Completely removes ZoneRental setup from a region.
     * This includes removing the sign from config and restoring the support block.
     *
     * @return true if sign was removed, false if no sign existed
     */
    fun removeRegionSetup(regionName: String, world: World): Boolean {
        val location = plugin.signsConfig.getSignLocation(regionName, world) ?: return false

        // Restore the support block if it exists
        if (plugin.signsConfig.hasSupportBlock(regionName, world)) {
            val supportLoc = plugin.signsConfig.getSupportBlockLocation(regionName, world)
            val supportData = plugin.signsConfig.getSupportBlockData(regionName, world)

            if (supportLoc != null && supportData != null) {
                try {
                    val supportBlock = supportLoc.getBlock()
                    val originalType = Material.valueOf(supportData["type"] ?: "")
                    val originalData = supportData["data"]

                    // Restore the block type
                    supportBlock.type = originalType

                    // Restore the block data if it exists
                    if (!originalData.isNullOrEmpty()) {
                        try {
                            val blockData = plugin.server.createBlockData(originalData)
                            supportBlock.blockData = blockData
                        } catch (e: IllegalArgumentException) {
                            plugin.logger.warning("Could not restore block data for support block: ${e.message}")
                        }
                    }

                    plugin.logger.info(
                        "Restored support block for region ${world.name}:$regionName to ${originalType.name}"
                    )
                } catch (e: IllegalArgumentException) {
                    plugin.logger.warning(
                        "Could not restore support block for region ${world.name}:$regionName: " +
                        "Invalid material type - ${e.message}"
                    )
                }
            }
        } else {
            // No support block data - manually break the sign
            val block = location.getBlock()
            if (block.state is Sign) {
                block.type = Material.AIR
            }
        }

        // Remove from config (this removes both sign and support block data)
        plugin.signsConfig.removeSign(regionName, world)
        plugin.logger.info("Removed ZoneRental setup from region ${world.name}:$regionName")

        return true
    }

    fun updateSign(regionName: String, world: World) {
        val location = plugin.signsConfig.getSignLocation(regionName, world) ?: return

        val block = location.getBlock()

        // Check if block is a sign
        val sign = block.state as? Sign
        if (sign == null) {
            plugin.logger.warning("Sign location for ${world.name}:$regionName is not a sign!")
            return
        }

        val rental = plugin.rentalManager.getRental(regionName, world)

        if (rental == null) {
            // Region is available
            updateAvailableSign(sign, regionName)
        } else {
            // Region is rented
            updateRentedSign(sign, rental)
        }

        sign.update(true)
    }

    private fun updateAvailableSign(sign: Sign, regionName: String) {
        val world = sign.world
        val price = plugin.configManager.getPriceForRegion(regionName, world)
        val duration = plugin.configManager.getDurationForRegion(regionName, world)
        val formattedPrice = String.format(plugin.configManager.currencyFormat, price)

        val formatList = plugin.configManager.availableSignFormat

        // Null/empty check to prevent NPE
        if (formatList.isNullOrEmpty()) {
            plugin.logger.warning("Available sign format is null or empty for region $regionName")
            return
        }

        var line = 0
        for (format in formatList) {
            if (line >= 4) break

            val text = format
                .replace("{region}", regionName)
                .replace("{price}", formattedPrice)
                .replace("{duration}", duration.toString())

            sign.line(line, text.toComponent())
            line++
        }
    }

    private fun updateRentedSign(sign: Sign, rental: Rental) {
        val formatList = plugin.configManager.rentedSignFormat

        // Null/empty check to prevent NPE
        if (formatList.isNullOrEmpty()) {
            plugin.logger.warning("Rented sign format is null or empty for region ${rental.regionName}")
            return
        }

        var line = 0
        for (format in formatList) {
            if (line >= 4) break

            val text = format
                .replace("{region}", rental.regionName)
                .replace("{owner}", rental.playerName)
                .replace("{expires}", rental.formattedEndDate)
                .replace("{days}", rental.daysRemaining.toString())
                .replace("{hours}", rental.hoursRemaining.toString())

            sign.line(line, text.toComponent())
            line++
        }
    }

    /**
     * Marks a sign as dirty (needs update).
     * Uses composite key format "world:region".
     */
    fun markSignDirty(regionName: String, world: World) {
        val compositeKey = "${world.name}:$regionName"
        dirtySigns.add(compositeKey)
    }

    /**
     * Marks multiple signs as dirty using composite keys (world:region format).
     * Used for bulk operations like group overrides.
     */
    fun bulkMarkSignsDirty(compositeKeys: List<String>) {
        dirtySigns.addAll(compositeKeys)
    }

    /**
     * Updates all dirty signs and clears the dirty set.
     * This is much more efficient than updating ALL signs every 30 seconds.
     */
    fun updateAllSigns() {
        if (dirtySigns.isEmpty()) {
            return // Nothing to update
        }

        val allSigns = plugin.signsConfig.getAllSigns()

        // Update only dirty signs
        for (compositeKey in dirtySigns) {
            val location = allSigns[compositeKey] ?: continue // Sign was removed

            // Parse composite key "world:region"
            val regionName = WorldRegionParser.extractRegionName(compositeKey) ?: continue
            val world = location.world ?: continue

            updateSign(regionName, world)
        }

        // Clear dirty set after updating
        dirtySigns.clear()
    }

    fun isRentalSign(location: Location): Boolean {
        return plugin.signsConfig.getRegionByLocation(location) != null
    }

    fun getRegionFromSign(location: Location): String? {
        return plugin.signsConfig.getRegionByLocation(location)
    }

    /**
     * Gets the region name if the given location is a protected support block.
     * Uses O(1) index lookup instead of O(n) linear scan for better performance.
     *
     * @return Composite key "world:region" if this is a support block, null otherwise
     */
    fun getSupportBlockRegion(location: Location): String? {
        // O(1) lookup via support block index
        return plugin.signsConfig.getSupportBlockByLocation(location)
    }
}
